cleverhans/attacks_tf.py

```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
+                            [-1, nb_features])
 
-    # Initialize the loop variables
-    iteration = 0
-    adv_x_original_shape = np.reshape(adv_x, original_shape)
-    current = utils_tf.model_argmax(sess, x, predictions, adv_x_original_shape,
-                                    feed=feed)
+    def condition(x_in, y_in, domain_in, i_in, cond_in):
+        # Repeat the loop until we have achieved misclassification or
+        # reaches the maximum iterations
+        return tf.logical_and(tf.less(i_in, max_iters), cond_in)
 
-    _logger.debug("Starting JSMA attack up to {} iterations".format(max_iters))
-    # Repeat this main loop until we have achieved misclassification
-    while (current != target and iteration < max_iters and
-           len(search_domain) > 1):
-        # Reshape the adversarial example
-        adv_x_original_shape = np.reshape(adv_x, original_shape)
+    def body(x_in, y_in, domain_in, i_in, cond_in):
+
+        preds = model.get_probs(x_in)
+        preds_onehot = tf.one_hot(tf.argmax(preds, axis=1), depth=nb_classes)
+
+        # create the Jacobian graph
+        list_derivatives = []
+        for class_ind in xrange(nb_classes):
+            derivatives = tf.gradients(preds[:, class_ind], x_in)
+            list_derivatives.append(derivatives[0])
+        grads = tf.reshape(tf.stack(list_derivatives),
+                           shape=[nb_classes, -1, nb_features])
 
         # Compute the Jacobian components
-        grads_target, grads_others = jacobian(sess, x, grads, target,
-                                              adv_x_original_shape,
-                                              nb_features, nb_classes,
-                                              feed=feed)
-
-        if iteration % ((max_iters + 1) // 5) == 0 and iteration > 0:
-            _logger.debug("Iteration {} of {}".format(iteration,
-                                                      int(max_iters)))
-        # Compute the saliency map for each of our target classes
-        # and return the two best candidate features for perturbation
-        i, j, search_domain = saliency_map(
-            grads_target, grads_others, search_domain, increase)
-
-        # Apply the perturbation to the two input features selected previously
-        adv_x = apply_perturbations(
-            i, j, adv_x, increase, theta, clip_min, clip_max)
-
-        # Update our current prediction by querying the model
-        current = utils_tf.model_argmax(sess, x, predictions,
-                                        adv_x_original_shape, feed=feed)
+        target_class = tf.reshape(tf.transpose(y_in, perm=[1, 0]),
+                                  shape=[nb_classes, -1, 1])
+        other_classes = tf.cast(tf.not_equal(target_class, 1), tf.float32)
+
+        grads_target = tf.reduce_sum(grads * target_class, axis=0)
+        grads_other = tf.reduce_sum(grads * other_classes, axis=0)
+
+        # Remove the already-used input features from the search space
```

Comment:
Why not just set the values of `grads_target` and `grads_other` to 0 for the features in domain_in? These way they will not match the `scores_mask` conditions below.

Reply:
The algorithm is picking pairs of pixels/features, so setting one pixel/feature to 0 doesn't prevent it from being picked again later. That's why their values need to be set based on the maximum value from the gradients.